ByteChannelContentSource

gregw opened the pull request:
This PR introduces a `ByteChannelContentSource` which is needed for the spring framework integration.

It also introduces a few minor utilities: 
 + a `ByteBufferPool.Sized` class that is a `ByteBufferPool.Wrapper` with a configured size and type so a no args  `acquire()` method can be called.   This avoid the need to many classes to have a three fields: pool, size and direct and also can avoid constructors/methods with size and direct  
 + an `ExceptionUtil.run(Runnable, Consumer<Throwable> failure)`  that is a simple try catch and avoids many private methods being created to do just that.

gregw committed:
Removed random access

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/PathContentSource.java:
```diff
@@ -84,22 +89,30 @@ public long getLength()
+        return byteBufferPool.isDirect();
     }
 
     public void setUseDirectByteBuffers(boolean useDirectByteBuffers)
```
Should I deprecate the direct and buffer size setters in favour of the constructor taking a Sized pool?

lorban replied:
+1

lachlan-roberts commented:
@gregw I looked at the `ContentWriterIteratingCallback` from the spring PR, and it looks like we already have something very similar, `org.eclipse.jetty.io.IOResources.PathToSinkCopier` which is used from `IOResources#copy(Resource, Content.Sink, ByteBufferPool, int, boolean, long, long, Callback)`.

So considering we already have this do we really need the `ByteChannelContentSource` at all?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java:
```diff
@@ -92,6 +93,40 @@ public void clear()
+    /**
+     * A ByteBufferPool for a specific size and type of buffer
+     */
+    class Sized extends Wrapper
```
+1 to this, we do pass the (ByteBufferPool, direction, size) triplet too often and this is a nice helper.

sbordet replied:
This class should not be a `Wrapper` or in general should not be a `is-a` `ByteBufferPool`, it should delegate to a BBP.

It's confusing to use because you pass a size, but then you can call `acquire(int)` with any size.

So now you have to downcast it to really use it, and at that point you don't use any of the super-class features, so it should not extend from BBP, but delegate to it.

sbordet replied:
It should be called `Fixed` rather than `Sized`, or `FixedSize`.

sbordet replied:
I still think the introduction of this class is orthogonal to this PR, that it should not be an _is-a_ BBP but just delegate to BBP, and it causes behavior changes that may be unexpected in minor Jetty releases.
I'd rather introduce it in 12.1.x.

gregw replied:
This class follows the pattern of the ThreadPool.SizedThreadPool.  It is just a normal BBP with an extra no-args method to get a specifically sized pool.     The name also follows the pattern from ThreadPool.    This doesn't introduce any changed behaviours, however some attempts to widely use other parts of this PR have caused changes. I'll back those out.  But I think this is OK as is.

gregw replied:
It is also a pattern used by VirtualThreadPool and TryExecutor, both of which add methods to their base-type, but still implement their base-type.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,221 @@
+ * <p>A {@link Content.Source} backed by a  {@link ByteChannel}.
+ * Any calls to {@link #demand(Runnable)} are immediately satisfied.</p>
+ */
+public class ByteChannelContentSource implements Content.Source
```
This looks very similar to `PathContentSource` and its private subclass `IOResources.RangedInputStreamContentSource`.

These 3 content sources should be rationalized.

gregw replied:
Maybe... but we need this in 12.0 and rationalizing these APIs sounds like more of a 12.1 thing

gregw replied:
Also, there are further improvements that can be made the RBB pull request

lorban replied:
I was talking about the implementation, `RangedInputStreamContentSource` is 90%+ identical to `ByteChannelContentSource`.

`PathContentSource` and `RangedInputStreamContentSource` could be made subclasses of `ByteChannelContentSource` with little effort, couldn't they?

gregw replied:
ok, I'll give it a try tomorrow

gregw replied:
@lorban I've cleaned up a bit of duplication... but I could keep going forever.... I think we should defer more until 21.1

gregw replied:
Hmmm as I feared, the cleanups have broken a bunch of tests.....  still working on this.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/InputStreamContentSource.java:
```diff
@@ -39,44 +40,55 @@ public class InputStreamContentSource implements Content.Source
-    private final ByteBufferPool bufferPool;
-    private int bufferSize = 4096;
-    private boolean useDirectByteBuffers;
+    private ByteBufferPool.Sized bufferPool;
```
-1 to this: if the buffer pool isn't final anymore, every access to that field should happen under a lock, or it should be made volatile.

I'd rather have a `setSize()` method in `ByteBufferPool.Sized` and keep this field final.

gregw replied:
This is tough... if we move a setSize to the Sized class, then those fields cannot be final there and that will persist forever....   However, if we live with a non final pool here, it is not going to be a problem as the setters are only used in an initialisation phase and when we ultimately remove the deprecated setters, then we can make this final again.

lorban replied:
I can buy this argument.

sbordet replied:
Not keep either, I think it's a 12.1.x change.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/PathContentSource.java:
```diff
@@ -40,20 +42,23 @@ public class PathContentSource implements Content.Source
-    private final ByteBufferPool byteBufferPool;
-    private int bufferSize = 4096;
-    private boolean useDirectByteBuffers = true;
+    private ByteBufferPool.Sized byteBufferPool;
```
Ditto: making this field non-final smells like a bad idea.

gregw replied:
Ditto - can we smell bad until we remove the setters in 12.1?

lorban replied:
OK.

lorban commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -65,7 +70,35 @@
+    private static ArrayByteBufferPool.Tracking byteBufferPool;
+
+    @BeforeEach
+    public void beforeAll()
```
This method is mis-named: `beforeAll()` vs the `@BeforeEach` annotation.

lorban commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ContentSourceTest.java:
```diff
@@ -65,7 +70,35 @@
+    }
+
+    @AfterEach
+    public void afterAll()
```
This method is mis-named: `afterAll()` vs the `@AfterEach` annotation.

gregw committed:
updates from review

gregw committed:
updates from review

gregw committed:
updates from review

gregw committed:
updates from review

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-ByteChannelContentSource

gregw committed:
updates from review

gregw committed:
updates from review

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathRequestContent.java:
```diff
@@ -62,4 +71,20 @@ public String getContentType()
+    /**
+     * @deprecated Use the {@link ByteBufferPool.Sized} in a constructor
+     */
+    @Deprecated (forRemoval = true)
```
Add the `since` attribute.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,303 @@
+    public ByteChannelContentSource(ByteBufferPool.Sized byteBufferPool, SeekableByteChannel seekableByteChannel, long offset, long length)
+    {
+        this(byteBufferPool, (ByteChannel)seekableByteChannel, offset, length);
+        if (offset > 0 && seekableByteChannel != null)
```
We do not know if we receive a `seekableByteChannel` with `position=0`, so I think we should enforce it anyway:

```suggestion
        if (offset >= 0 && seekableByteChannel != null)
```

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,303 @@
+            ExceptionUtil.run(demandCallback, this::fail);
+    }
+
+    private void checkOpenLocked()
```
Elsewhere we always use `locked` as prefix, so rename to `lockedCheckOpen()`.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,303 @@
+            {
+                _buffer = _byteBufferPool.acquire();
+            }
+            else if (_buffer.isRetained())
```
I don't think we should try to reuse the `_buffer`, as it would be in a race with possible usages by the application.

The application reads, then gives the buffer to some other thread to be written, assuming there is no need to retain because now the application is in charge or `release()`, then calls read again, and here we release it and overwrite the bytes.

gregw replied:
This is not an issue.  once the app releases the buffer, then they cannot re-retain it. If they don't release, we don't reuse.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,303 @@
+        {
+            checkOpenLocked();
+            if (_terminal != null || _byteChannel == null || !_byteChannel.isOpen())
+                return false;
```
That seems wrong.

After reading from a file, the `_byteChannel` is closed, but we can certainly rewind and read from it again.

gregw replied:
i've changed checkopen to be able to remove terminal

sbordet replied:
I think this implementation is too complicated.
It should just reset all the relevant fields, and rely on the fact that the first call to `read()` will lazily open the channel.
This will avoid code duplication like `instanceof SeekableByteChannel` and setting the position.

See the original implementation in `PathContentSource`.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/InputStreamContentSource.java:
```diff
@@ -39,44 +40,65 @@ public class InputStreamContentSource implements Content.Source
+     * @param useDirectByteBuffers {@code true} if direct buffers will be used.
+     * @deprecated Use {@link InputStreamContentSource#InputStreamContentSource(InputStream, ByteBufferPool.Sized)}
+     */
+    @Deprecated(forRemoval = true)
```
Missing `since` attribute.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/PathContentSource.java:
```diff
@@ -22,38 +23,45 @@
  * <p>A {@link Content.Source} that provides the file content of the passed {@link Path}.</p>
+ * @deprecated use {@link ByteChannelContentSource.PathContentSource}
  */
+@Deprecated(forRemoval = true)
 public class PathContentSource implements Content.Source
```
Why could not this class remain un-deprecated, and extend `ByteChannelContentSource`?

gregw replied:
@sbordet because this class is configured with setters and the new one uses constructor injection

sbordet replied:
It is a breaking behavior change then, not optimal for a dot release.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java:
```diff
@@ -46,43 +47,45 @@ public InvocationType getInvocationType()
+            current = next;
+            failed(failure);
+            return Action.SCHEDULED;
+        }
```
This seems more complicated than needs to be, I would just throw.

For example, current is a failure chunk, `next(current)` likely returns the chunk itself, we release the failure chunk, and then `onCompleteFailure()` releases it again.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/internal/ContentCopierTest.java:
```diff
@@ -16,21 +16,51 @@
 
 public class ContentCopierTest
 {
+
```
Empty line.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ExceptionUtil.java:
```diff
@@ -300,8 +303,102 @@ public static Throwable combine(Throwable t1, Throwable t2)
     }
 
-    private ExceptionUtil()
+    public static void callThen(Throwable cause, Consumer<Throwable> first, Consumer<Throwable> second)
```
Rename to `callAndThen()`.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,303 @@
+            }
+            catch (Throwable t)
+            {
+                _terminal = Content.Chunk.from(t, true);
```
Must release the buffer here.
Also close the `_byteChannel`?

gregw committed:
updates from review

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,304 @@
+            ExceptionUtil.run(demandCallback, this::fail);
+    }
+
+    private void lockedCheckOpen()
+    {
+        if (_terminal == null || !Content.Chunk.isFailure(_terminal) && _terminal.isLast())
+        {
+            _terminal = null;
+            if (_byteChannel == null || !_byteChannel.isOpen())
+            {
+                try
+                {
+                    _byteChannel = open();
+                    if (_byteChannel == null || !_byteChannel.isOpen())
+                        _terminal = Content.Chunk.from(new ClosedChannelException(), true);
+                    if (_offset > 0 && _byteChannel instanceof SeekableByteChannel seekableByteChannel)
+                        seekableByteChannel.position(_offset);
+                }
+                catch (IOException e)
+                {
+                    _terminal = Content.Chunk.from(e, true);
+                }
+
+                if (_terminal != null && _buffer != null)
+                {
+                    _buffer.release();
+                    _buffer = null;
+                }
+            }
+        }
+    }
```
I think this method does way too much, too many side effects calling it.

I think there should be a method used from `read()` that opens the channel if possible, and a different method to be used from `rewind()`.

For example right now: app reads, then reads EOF; if the app insist reading it, it should receive another EOF, but instead `lockedCheckOpen()` will clear _terminal and re-open the channel -- not good.

I feel this class should inherit the simplicity of `PathContentSource` rather than the current more complex implementation.

gregw replied:
Well this class was simple, but @lorban asked for it to be used to replace many other similar classes, so it is now a bit more complex.   Ultimately this class should be able to replace 3 or 4 other classes once some small behaviour changes are permitted in 12.1.

However, I will remove the rewind logic from this method.

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathRequestContent.java:
```diff
@@ -26,7 +26,7 @@
  * to read from the {@code Path}.</p>
  */
-public class PathRequestContent extends PathContentSource implements Request.Content
+public class PathRequestContent extends ByteChannelContentSource.PathContentSource implements Request.Content
```
This class should still inherit from `io.content.PathContentSource`, which itself should inherit from the new `ByteChannelContentSource`.

gregw replied:
@sbordet I had initially deferred these changes to 12.1, but @lorban asked for them to be in this PR.  I will back the substantive changes out and simply deprecate the methods that will change in 12.1.x

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathRequestContent.java:
```diff
@@ -62,4 +71,20 @@ public String getContentType()
+    /**
+     * @deprecated Use the {@link ByteBufferPool.Sized} in a constructor
+     */
+    @Deprecated (forRemoval = true, since = "12.0.11")
+    public void setUseDirectByteBuffers()
+    {
+    }
```
This is now a change in behavior:

```java
var s = new PathContentSource(path);
s.setBufferSize(10 * 1024);
```

Updating Jetty results in the above code to compile, but not work as expected.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,304 @@
+    /**
+     * A {@link ByteChannelContentSource} for a {@link Path}
+     */
+    public static class PathContentSource extends ByteChannelContentSource
```
Should not be an inner class, but rather be retrofitted into the existing top-level `PathContentSource`.

gregw replied:
That can't be done without changes in behaviour.  So will implement here for now, but as deprecated.  In 12.1 the PathContentSource can be changed to take this implementation directly

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java:
```diff
@@ -46,43 +47,43 @@ public InvocationType getInvocationType()
     @Override
     protected Action process() throws Throwable
     {
+        if (current != null)
+            current.release();
```
I think this is better done in an override of `succeeded()`, which BTW would not require the null check.

gregw replied:
@sbordet, as I keep saying (in other PRs), overriding succeeded and messing with fields there is an anti-pattern.  ICB enforces the concurrency that the process method is only ever called by a single thread.

A field that is changed in succeeded may be altered by the a call made from process, it may be done in the same thread, or a different one.  There is no memory boundary and the behaviour will be undefined.

The whole point of ICB is to serialize events into the process method, which can determine a specific ordering of them.   Recycling buffers that might be in use in another method is just not good practise, even if we "know" that the process method doesn't use them after scheduling an action.

gregw committed:
updates from review

Reverted PathContentSource to previous implementation, but with deprecated methods
Simplified ByteChannelContentSource
more rewind tests.

gregw commented:
@lorban I have met @sbordet's review as much as I can.   Can you please re-review this ASAP!

gregw committed:
updates from review

Reverted PathContentSource to previous implementation, but with deprecated methods
Simplified ByteChannelContentSource
more rewind tests.

gregw commented:
@lachlan-roberts can  you review this in light of @sbordet being on vacation.  Can  you check if I have addressed all his points sufficiently.
Note that @lorban has previously given this the green tick, before the changes requested by @sbordet (but didn't re-review yesterday).

lachlan-roberts commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPart.java:
```diff
@@ -475,7 +475,8 @@ public Path getPath()
         public Content.Source newContentSource()
         {
-            return new PathContentSource(getPath());
+            // TODO: use a ByteBuffer pool and direct ByteBuffers?
```
Do you plan to resolve this TODO in this PR?

gregw replied:
I had not.  This TODO is inherited from existing implementation.  I wasn't really even planning on extensively using the new class until 12.1, but was pushed to do so in review.  I really just want one little class added without a huge revolution in 12.0.  This PR is already too big for that intention, so I'd rather leave this TODO for 12.1

lachlan-roberts commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathRequestContent.java:
```diff
@@ -47,7 +47,7 @@ public PathRequestContent(String contentType, Path filePath) throws IOException
     {
-        this(contentType, filePath, null);
+        this(contentType, filePath, (ByteBufferPool)null);
         setBufferSize(bufferSize);
```
This method is deprecated, can you use the super constructor which takes the size instead.

lachlan-roberts commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/PathContentSource.java:
```diff
@@ -36,24 +38,39 @@
+        this(path, new ByteBufferPool.Sized(byteBufferPool, true, -1));
+    }
+
+    public PathContentSource(Path path, ByteBufferPool.Sized byteBufferPool)
```
I don't think this works as you expect, for example if you do:
```
ByteBufferPool bufferPool = new ByteBufferPool.Sized(new ArrayByteBufferPool());
new PathContentSource(path, bufferPool);
```

then it ends up in the other constructor and re-wraps and overrides the values from the sized pool. So maybe you need an `instanceOf` check for `ByteBufferPool.Sized` in the above constructor?

lachlan-roberts commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,297 @@
+
+        public PathContentSource(ByteBufferPool.Sized byteBufferPool, Path path)
+        {
+            super(byteBufferPool, null, 0, size(path));
```
Not sure the `size(path)` is necessary, can't we just do `length=-1` and read to EOF and save the extra fileSystem access to look up the file size?

gregw replied:
The previous impl did that, but I don't see us saving much, other than for tiny files.  So OK.

gregw replied:
Turns out that it is needed by the multipart tests that expect to known how long a part is.  reverting the change.

lachlan-roberts commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java:
```diff
@@ -0,0 +1,297 @@
+     * A {@link ByteChannelContentSource} for a {@link Path}
+     * @deprecated To be replaced by an updated {@link org.eclipse.jetty.io.content.PathContentSource} in 12.1.0
+     */
+    @Deprecated(forRemoval = true, since = "12.0.11")
```
Weird that this is already deprecated without a replacement available, and is still used in many places in the code.

gregw replied:
yeah.... well I've been pushed in both directions by other review.  Ultimately we don't want to change the behavior on the other `PathContentSource` changed in a dot release, but we want to establish the new class with better semantic now.  It is just on notice that it will move to the other location in 12.1.  What else can we do to introduce a new semantic but not change the old until a point release.

gregw committed:
updates from review

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-ByteChannelContentSource

gregw committed:
reverted updates from review
